File: tests/python/test_spatial_prioritization.py
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

"""Tests for spatial-prioritization skill scripts.
Covers: run_prioritization.R and prioritization_sensitivity.R logic.
"""
import pytest


def test_target_must_be_in_valid_range():
    """Conservation targets must be between 0 and 1."""
    targets = [0.17, 0.30, 0.40, 0.50, 0.60]
    for t in targets:
        assert 0 < t <= 1.0


def test_iucn_targets_ordering():
    """IUCN-based targets: CR > EN > VU > NT > LC."""
    iucn_targets = {
        "CR": 0.60,
        "EN": 0.50,
        "VU": 0.40,
        "NT": 0.30,
        "LC": 0.17,
    }
    assert iucn_targets["CR"] > iucn_targets["EN"]
    assert iucn_targets["EN"] > iucn_targets["VU"]
    assert iucn_targets["VU"] > iucn_targets["NT"]
    assert iucn_targets["NT"] > iucn_targets["LC"]


def test_loglinear_target_at_boundary():
    """Log-linear target for very large species should not fall below lower bound."""
    def loglinear_target(range_km2, lower=0.10, upper=1.0,
                         min_area=1.0, max_area=250000.0):
        import math
        if range_km2 <= min_area:
            return upper
        if range_km2 >= max_area:
            return lower
        frac = (math.log(range_km2) - math.log(min_area)) / \
               (math.log(max_area) - math.log(min_area))
        return upper - frac * (upper - lower)

    t_large = loglinear_target(250000.0)
    t_small = loglinear_target(1.0)
    assert abs(t_large - 0.10) < 1e-9
    assert abs(t_small - 1.00) < 1e-9


def test_feature_representation_is_nonnegative():
    """Feature representation (% of target met) must be ≥ 0."""
    representation = [
        {"feature": "Jaguar",        "held": 0.32, "target": 0.30, "met": True},
        {"feature": "Giant anteater","held": 0.28, "target": 0.30, "met": False},
        {"feature": "Tapir",         "held": 0.31, "target": 0.30, "met": True},
    ]
    for feat in representation:
        assert feat["held"] >= 0
        assert feat["target"] >= 0


def test_blm_calibration_has_required_columns():
    """BLM calibration output must include blm, total_cost, total_boundary columns."""
    calibration = [
        {"blm": 0.0,   "total_cost": 1245.3, "total_boundary": 3820.1, "n_pu": 214},
        {"blm": 0.001, "total_cost": 1312.8, "total_boundary": 2941.6, "n_pu": 218},
        {"blm": 0.01,  "total_cost": 1489.2, "total_boundary": 2215.0, "n_pu": 227},
    ]
    required = {"blm", "total_cost", "total_boundary"}
    for row in calibration:
        assert required.issubset(set(row.keys()))

    # Cost should be non-decreasing with increasing BLM
    costs = [r["total_cost"] for r in calibration]
    assert all(costs[i] <= costs[i + 1] for i in range(len(costs) - 1))


def test_irreplaceability_in_valid_range():
    """Irreplaceability scores must be in [0, 1]."""
    irreplaceability = [0.0, 0.12, 0.45, 0.87, 1.0]
    assert all(0 <= v <= 1 for v in irreplaceability)


def test_portfolio_frequency_in_valid_range():
    """Portfolio selection frequency across scenarios must be in [0, 1]."""
    # Frequency = how many scenarios selected this planning unit / total scenarios
    n_scenarios = 10
    frequencies = [3, 7, 10, 0, 5]
    for f in frequencies:
        freq_norm = f / n_scenarios
        assert 0 <= freq_norm <= 1

File: tests/r/test-spatial-prioritization.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# Tests for spatial-prioritization skill scripts
# Covers: run_prioritization.R, prioritization_sensitivity.R
library(testthat)

test_that("representation targets are in valid range [0, 1]", {
  targets <- c(0.17, 0.30, 0.40, 0.50, 0.60)
  expect_true(all(targets >= 0 & targets <= 1))
})

test_that("IUCN-based targets are higher for threatened species", {
  iucn_targets <- c(CR = 0.60, EN = 0.50, VU = 0.40, NT = 0.30, LC = 0.17)
  expect_gt(iucn_targets["CR"], iucn_targets["EN"])
  expect_gt(iucn_targets["EN"], iucn_targets["VU"])
  expect_gt(iucn_targets["VU"], iucn_targets["NT"])
})

test_that("log-linear target returns 1.0 for species below lower threshold", {
  range_km2 <- 500  # below 1000 km2 lower threshold
  lower_threshold <- 1000
  upper_threshold <- 250000
  lower_target <- 0.10
  upper_target <- 1.00
  target <- ifelse(range_km2 < lower_threshold, upper_target,
            ifelse(range_km2 > upper_threshold, lower_target,
                   upper_target + (lower_target - upper_target) *
                     (log(range_km2) - log(lower_threshold)) /
                     (log(upper_threshold) - log(lower_threshold))))
  expect_equal(target, 1.00)
})

test_that("feature representation check computes correctly", {
  rep_df <- data.frame(
    feature      = c("sp_a", "sp_b", "sp_c"),
    relative_held = c(0.35, 0.22, 0.28),
    target       = c(0.30, 0.30, 0.30)
  )
  rep_df$target_met <- rep_df$relative_held >= rep_df$target
  expect_equal(sum(rep_df$target_met), 2L)   # sp_a and sp_c meet target
  expect_false(rep_df$target_met[2])          # sp_b does not meet target
})

test_that("BLM calibration data has required columns", {
  blm_df <- data.frame(
    blm        = c(0, 0.01, 0.1, 1.0),
    cost       = c(234, 251, 312, 498),
    boundary   = c(18240, 11880, 7200, 4320),
    n_selected = c(487, 508, 565, 621)
  )
  expect_true(all(c("blm", "cost", "boundary") %in% names(blm_df)))
  # Cost increases with BLM
  expect_true(all(diff(blm_df$cost) > 0))
  # Boundary decreases with BLM
  expect_true(all(diff(blm_df$boundary) < 0))
})

test_that("irreplaceability values are in [0, 1]", {
  irr_vals <- c(0.0, 0.25, 0.72, 0.95, 1.0)
  expect_true(all(irr_vals >= 0 & irr_vals <= 1))
})
